### PR TITLE
feat(dsa): add SM80 Torch fallbacks and Triton paged-MQA indexer

### PR DESCRIPTION
**Pull request by xiangui33423**
## Motivation

  The current CUDA DSA path cannot run end-to-end on NVIDIA A100/SM80:

  1. The paged-MQA indexer depends on a DeepGEMM FP8 path targeting newer architectures.
  2. Sparse MLA attention uses an FA3 configuration with asymmetric head dimensions (`QK=576`, `V=512`) that is unsupported on SM80.

  This PR establishes an explicit SM80 DSA path in two steps:

  - Correctness-first Torch backends for paged-MQA logits and sparse-MLA attention.
  - An optimized native Triton paged-MQA indexer for A100.

  The attention path remains on the Torch fallback in this PR. Triton sparse-MLA attention is intentionally left for a follow-up.

  All new backends are explicit opt-in and existing automatic backend selection remains unchanged.

  ## Modifications

  ### Torch correctness backends

  Add a general Torch paged-MQA implementation for the packed DSA index cache:

  - Query: `[T, H, 128]` E4M3FN
  - Packed page: `[64 * 128 FP8 key bytes][64 FP32 scales]`
  - Output: `[T, max_seq_len]` FP32
  - Query/page chunking avoids gathering the complete cache.
  - Invalid pages and positions outside the sequence length produce deterministic `0.0`.

  Add a Torch sparse-MLA implementation:

  - BF16 query and KV cache
  - FP32 score computation and online softmax state
  - Query and Top-K chunking
  - Correct handling of `-1` sparse indices
  - No complete `[T, H, K, 576]` intermediate tensor

  These implementations provide the correctness reference and fallback for the optimized SM80 path.

  ### SM80 Triton paged-MQA indexer

  Add a native Triton paged-MQA logits kernel for NVIDIA A100:

  - Fixed grid: `(query row, logical KV page)`
  - One program computes 64 logits for one query and one cache page.
  - Query and key tensors are passed as `uint8` views.
  - E4M3FN is decoded in software without using an SM80-unsupported Triton FP8 type or FP8 MMA.
  - Decoded values are converted to BF16 for SM80 tensor-core `tl.dot`.
  - Dot products, ReLU, head-weight reduction, and final accumulation use FP32.
  - The implementation uses 16-head tiles, 4 warps, and 1 pipeline stage.
  - Only the full FP32 logits output is allocated; cache scales and FP8 bytes are accessed through views.

  E4M3FN handling explicitly covers:

  - Positive and negative zero
  - Subnormal and normal values
  - Maximum finite values
  - `0x7f` and `0xff` NaN patterns

  NaN patterns propagate deterministically instead of being decoded as finite values.

  ### Backend integration

  Add:

  ```bash
  --dsa-paged-mqa-logits-backend triton

  The Triton path:

  - Supports NVIDIA compute capability 8.0 only.
  - Rejects ROCm, SM86/SM89, Hopper, and Blackwell with a clear error.
  - Supports ordinary eager decode with multiple query rows.
  - Handles q_offset and restores -1 results for padding query rows.
  - Supports non-contiguous and repeated physical pages.
  - Writes strict 0.0 for negative/out-of-range pages and invalid token positions.
  - Does not build, refresh, or call DeepGEMM paged-MQA metadata or kernels.
  - Explicitly rejects target verification and draft-extend-v2, with instructions to use the Torch indexer fallback.

  The Torch path continues to be available as:

  --dsa-paged-mqa-logits-backend torch

  ### Runtime constraints

  The first SM80 configuration requires:

  - NVIDIA A100/SM80
  - BF16 main KV cache
  - CUDA Graph disabled
  - Overlap scheduling disabled
  - Compatible legacy SGL Top-K instead of the Hopper cluster-based Top-K v2
  - Torch sparse-MLA attention

  Example:

  python -m sglang.launch_server \
    --model-path MODEL_PATH \
    --tp TP_SIZE \
    --kv-cache-dtype bfloat16 \
    --dsa-paged-mqa-logits-backend triton \
    --dsa-prefill-backend torch \
    --dsa-decode-backend torch \
    --disable-cuda-graph

  ## Accuracy Tests

  Test environment:

  GPU:      NVIDIA A100 80GB PCIe
  Compute:  SM80 / compute capability 8.0
  PyTorch:  2.13.0+cu130
  CUDA:     13.0
  Triton:   3.7.1

  Coverage includes:

  - Exhaustive testing of all 256 E4M3FN byte patterns
  - Comparison with an independent Python bit-level decoder
  - Comparison with PyTorch float8_e4m3fn -> float32
  - H = 8, 16, 32, 64
  - T = 1, 2, 8, 32
  - Sequence lengths 1, 63, 64, 65, and 4096
  - Non-64-aligned max_seq_len
  - Non-contiguous and repeated page tables
  - Negative and out-of-range physical pages
  - Different sequence lengths per query row
  - Padding query rows
  - Maximum finite, subnormal, zero, and mixed-sign FP8 values
  - Strict zero output for invalid positions
  - Non-tie Top-K equality
  - Mock verification that the Triton and Torch paths never call:
      - deep_gemm.get_paged_mqa_logits_metadata
      - deep_gemm.fp8_paged_mqa_logits

  - Ordinary sparse-MLA prefill with the Torch attention backend
  - 32 consecutive eager decode steps using:
      - Triton paged-MQA indexer
      - Compatible SGL/Torch Top-K
      - Torch sparse-MLA attention

  Final regression result:

  32 passed, 2 skipped

  The two skipped tests require H100/H200 for DeepGEMM and TileLang cross-backend golden generation.

  A full model end-to-end test was not run because the target checkpoint and deployment configuration were not available in the workspace.

  ## Speed Tests and Profiling

  Benchmark methodology:

  - Synchronize before first-call timing.
  - Record first-call/JIT latency separately.
  - Warm up Triton 5 times and Torch 2 times.
  - Measure steady-state latency with CUDA events.
  - Repeat Triton 20 times and Torch 3 times.
  - Measure peak allocation relative to the live-input memory baseline.

  Representative H=64 results:

     T    Context         Torch      Triton    Speedup
  ━━━━━  ━━━━━━━━━  ━━━━━━━━━━━━  ━━━━━━━━━━  ━━━━━━━━━
     8         4K      2.076 ms    0.065 ms     31.88x
  ─────  ─────────  ────────────  ──────────  ─────────
     8        32K     15.597 ms    0.405 ms     38.53x
  ─────  ─────────  ────────────  ──────────  ─────────
    32        32K     62.830 ms    1.576 ms     39.87x
  ─────  ─────────  ────────────  ──────────  ─────────
     8       128K     61.293 ms    1.576 ms     38.89x
  ─────  ─────────  ────────────  ──────────  ─────────
    32       128K    255.605 ms    6.248 ms     40.91x

  The acceptance matrix covers:

  - T = 1, 8, 32
  - H = 8, 16, 32, 64
  - Context lengths 4K, 32K, and 128K

  The minimum representative long-context speedup is 38.53x, exceeding the initial 5x target.

  Observed cold Triton compilation time for a new specialization is approximately 0.77–1.64 seconds. No additional GPU allocation beyond the
  complete FP32 logits output was measured for the Triton path.

  ## Scope and follow-ups

  Not included in this PR:

  - Triton sparse-MLA attention
  - CUDA Graph capture/replay
  - Speculative target verification
  - Draft extend v2
  - Context parallelism
  - PD disaggregation or HiCache
  - Triton replacement for the ragged prefill indexer
  - Automatic SM80 backend selection
  - SM86/SM89 support
  - Full model end-to-end validation
  - H100/H200 cross-backend golden execution

  ## Checklist

  - [x] Format code with the repository pre-commit configuration.
  - [x] Add unit, kernel, integration, and benchmark tests.
  - [x] Validate on an actual NVIDIA A100/SM80 GPU.
  - [x] Provide correctness and performance results.
  - [x] Preserve existing backend defaults.
  - [x] Keep the new SM80 backends explicit opt-in.
  - [x] Verify that Torch/Triton paths bypass DeepGEMM paged-MQA APIs.
  - [ ] Run H100/H200 cross-backend golden tests on matching hardware.
  - [ ] Run a full model end-to-end accuracy test when the target checkpoint is available.

<!-- pr-states:start -->
---
### CI States

Latest PR Test (Base): <!-- slot:pr-test:start -->:x: [Run #32209113918](https://github.com/sgl-project/sglang/actions/runs/32209113918)<!-- slot:pr-test:end -->
Latest PR Test (Extra): <!-- slot:pr-test-extra:start -->:x: [Run #32209113784](https://github.com/sgl-project/sglang/actions/runs/32209113784)<!-- slot:pr-test-extra:end -->
<!-- pr-states:end -->

### CODE REVIEW

**Comment by xiangui33423** [comment]
#35420
